src/Compilers/CSharp/Test/Emit3/Semantics/PatternMatchingTests4.cs

```diff
@@ -4829,5 +4829,988 @@ class D : A { }
+                // (6,24): error CS9268: The pattern is redundant. Did you mean to parenthesize the disjunctive 'or' pattern?
+                // _ = b is (not null) or true; // 4
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "true").WithLocation(6, 24),
+                // (11,5): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
+                //     false => 43, // 5
+                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "false").WithLocation(11, 5),
+                // (17,5): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
+                //     true => 43, // 6
+                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "true").WithLocation(17, 5));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75506")]
+        public void RedundantPattern_NotNullOrBooleanLiteralPattern_WrongType()
+        {
+            var source = """
+#nullable enable
+int? i = null;
+_ = i is not null or false; // 1
+_ = i is (not null) or false; // 2
+
+_ = i switch
+{
+    not null => 42,
+    false => 43, // 3
+    _ => 44
+};
+""";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (3,22): error CS0029: Cannot implicitly convert type 'bool' to 'int?'
+                // _ = i is not null or false; // 1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "false").WithArguments("bool", "int?").WithLocation(3, 22),
+                // (3,22): error CS9268: The pattern is redundant. Did you mean to parenthesize the disjunctive 'or' pattern?
+                // _ = i is not null or false; // 1
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "false").WithLocation(3, 22),
+                // (4,24): error CS0029: Cannot implicitly convert type 'bool' to 'int?'
+                // _ = i is (not null) or false; // 2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "false").WithArguments("bool", "int?").WithLocation(4, 24),
+                // (4,24): error CS9268: The pattern is redundant. Did you mean to parenthesize the disjunctive 'or' pattern?
+                // _ = i is (not null) or false; // 2
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "false").WithLocation(4, 24),
+                // (9,5): error CS0029: Cannot implicitly convert type 'bool' to 'int?'
+                //     false => 43, // 3
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "false").WithArguments("bool", "int?").WithLocation(9, 5));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75506")]
+        public void RedundantPattern_NotNullOrEmptyListPattern()
+        {
+            var source = """
+#nullable enable
+int[]? i = null;
+_ = i is not null or []; // 1
+_ = i is (not null) or []; // 2
```

Comment:
It feels like it would be good for more of these tests to have the "happy case" as well, for example, demonstrating that `i is not (null or [])` does not warn.